molgenis-data-vcf/src/main/java/org/molgenis/data/vcf/VcfRepository.java

```diff
@@ -34,98 +33,118 @@
 	private final String entityTypeId;
 	private final VcfAttributes vcfAttributes;
 	private final EntityTypeFactory entityTypeFactory;
 	private final AttributeFactory attrMetaFactory;
-	protected final Supplier<VcfToEntity> vcfToEntitySupplier;
+	private VcfToEntity vcfToEntity;
+	private final File file;
 
-	public VcfRepository(File file, String entityTypeId, VcfAttributes vcfAttributes,
-			EntityTypeFactory entityTypeFactory, AttributeFactory attrMetaFactory)
+	VcfRepository(File file, String entityTypeId, VcfAttributes vcfAttributes, EntityTypeFactory entityTypeFactory,
+			AttributeFactory attrMetaFactory)
 	{
-		this(new VcfReaderFactoryImpl(file), entityTypeId, vcfAttributes, entityTypeFactory, attrMetaFactory);
-	}
-
-	protected VcfRepository(VcfReaderFactory vcfReaderFactory, String entityTypeId, VcfAttributes vcfAttributes,
-			EntityTypeFactory entityTypeFactory, AttributeFactory attrMetaFactory)
-	{
-		this.vcfReaderFactory = requireNonNull(vcfReaderFactory);
+		this.file = requireNonNull(file);
 		this.entityTypeId = requireNonNull(entityTypeId);
 		this.vcfAttributes = requireNonNull(vcfAttributes);
 		this.entityTypeFactory = requireNonNull(entityTypeFactory);
 		this.attrMetaFactory = requireNonNull(attrMetaFactory);
-		this.vcfToEntitySupplier = Suppliers.memoize(this::parseVcfMeta);
+		parseVcfMeta();
 	}
 
-	private VcfToEntity parseVcfMeta()
+	private void parseVcfMeta()
 	{
-		VcfReader reader = vcfReaderFactory.get();
-		try
-		{
-			VcfMeta vcfMeta = reader.getVcfMeta();
-			return new VcfToEntity(entityTypeId, vcfMeta, vcfAttributes, entityTypeFactory, attrMetaFactory);
-		}
-		catch (IOException | RuntimeException e)
-		{
-			throw new MolgenisDataException("Failed to read VCF Metadata from file", e);
-		}
-		finally
+		withReader(reader ->
 		{
 			try
 			{
-				reader.close();
+				VcfMeta vcfMeta = reader.getVcfMeta();
+				vcfToEntity = new VcfToEntity(entityTypeId, vcfMeta, vcfAttributes, entityTypeFactory, attrMetaFactory);
 			}
-			catch (IOException e)
+			catch (IOException | RuntimeException e)
 			{
-				LOG.info("Failed to close VcfReader", e);
+				throw new MolgenisDataException("Failed to read VCF Metadata from file", e);
 			}
-		}
+		});
 	}
 
-	/**
-	 * Returns an iterator for this repository.
-	 * <p>
-	 * Use with caution! Multiple iterators will all point to the same line in the VCF file, leading to unpredictable
-	 * behaviour. If you want to get the EntityType of this repository and you can't access getEntityType(),
-	 * convert the iterator to a PeekingIterator and peek the first Entity.
-	 */
 	@Override
 	public Iterator<Entity> iterator()
 	{
-		Iterator<VcfRecord> vcfRecordIterator = Iterators.unmodifiableIterator(vcfReaderFactory.get().iterator());
-		VcfToEntity vcfToEntity = vcfToEntitySupplier.get();
-		return Iterators.transform(vcfRecordIterator, vcfToEntity::toEntity);
+		throw new UnsupportedOperationException("Use forEachBatched instead of iterator");
```

Comment:
This violates the Repository contract. How about storing a Closable locally for each call to iterator() and closing the Closables on calling Repository.close?

Reply:
I (We?)'d like to get rid of the calls to iterator() altogether and for VcfRepository we checked that none are left after this PR. The whole VcfReaderRegistry is a bug waiting to happen so I'd say good riddance to it. Would introducing a RepositoryCapability.ITERABLE help keep the contract safe? How much work till we can get rid of the method altogether?

Reply:
There's [a WIP here](https://github.com/dennishendriksen/molgenis/tree/refactor/removeRepositoryIterator) to remove the Iterable contract from Repositories altogether. It's based on this PR. We can't be sure that we'll have time to finish it in the near future. 

We explored two options concerning this PR:
1. Merge as is. The ```iterator()``` will throw an ```UnsupportedOperationException()``` and we'll just have to make sure that nobody uses it. 
2. Suppress the Sonar warning for now and merge this fix later as part of the branch mentioned above. 

After some discussion we decided to go for option 1.